DataModel/SubSample.cpp

```diff
@@ -1,40 +1,130 @@
 #include "SubSample.h"
-
 #include <cassert>
 
-SubSample::SubSample(std::vector<int> PMTid,std::vector<float> time)
+SubSample::SubSample(std::vector<int> PMTid, std::vector<TimeDelta::short_time_t> time, std::vector<float> charge, TimeDelta timestamp)
 {
-  assert(PMTid.size() == time.size());
-  m_PMTid=PMTid;
-  m_time=time;
-  //fill the int versions
-  m_time_int.resize(m_time.size());
-  for(unsigned int i = 0; i < m_time.size(); i++) {
-    m_time_int[i] = m_time[i];
+  // If charge vector is empty, fill with 0s
+  if (charge.size() == 0){
+    charge =  std::vector<float>(PMTid.size(), 0.);
   }
-}
-
-SubSample::SubSample(std::vector<int> PMTid, std::vector<float> time, std::vector<float> charge)
-{
+  // Check that all vectors are same size
   assert(PMTid.size() == time.size() && PMTid.size() == charge.size());
+  // Assign values
   m_PMTid  = PMTid;
   m_time   = time;
   m_charge = charge;
-  //fill the int versions
-  m_time_int.resize(m_time.size());
-  m_charge_int.resize(m_time.size());
-  for(unsigned int i = 0; i < m_time.size(); i++) {
-    m_time_int[i] = m_time[i];
-    m_charge_int[i] = m_charge[i];
+  m_timestamp = timestamp;
+}
+
+void SubSample::SortByTime(){
+  //algorithm borrowed from WCSimWCDigi::SortArrayByHitTime()
+  int i, j;
+  TimeDelta::short_time_t save_time;
+  int save_PMTid;
+  double save_charge;
+
+  for (i = 1; i < m_PMTid.size(); ++i) {
+    save_time       = m_time[i];
+    save_PMTid      = m_PMTid[i];
+    save_charge     = m_charge[i];
+    for (j = i; j > 0 && m_time[j-1] > save_time; j--) {
+      m_time[j]     = m_time[j-1];
+      m_PMTid[j]    = m_PMTid[j-1];
+      m_charge[j]   = m_charge[j-1];
+    }//j
+    m_time[j]     = save_time;
+    m_PMTid[j]    = save_PMTid;
+    m_charge[j]   = save_charge;
+  }//i
+}
+
+bool SubSample::IsSortedByTime() const {
+  int size = m_time.size();
+  for (int i=0; i<size-1; ++i){
+    if (m_time[i] > m_time[i+1]){
+      return false;
+    }
+  }
+  return true;
+}
+
+std::vector<SubSample> SubSample::Split(TimeDelta target_width, TimeDelta target_overlap) const {
+
+  // If the sample is empty, just return a copy of self
+  if (m_time.size() == 0){
+    return std::vector<SubSample>(1, *this);
   }
+
+
+  // Distance between SubSamples
+  TimeDelta target_stride = target_width - target_overlap;
+
+  // Ensure everything is sorted
+  assert(IsSortedByTime());
```

Comment:
```suggestion
  if(!IsSortedByTime())
    SortByTime();
```
@brichards64 can comment, but the idea of the DAQ (and therefore same here in TriggerApp) is to be as good at catching its own errors and carrying on as possible.
Probably needs a comment in CONTRIBUTING alluding to this, and forbidding `assert()`

Reply:
That would make the method no longer being `const`, since it would potentially change the object. This might be unexpected behaviour when you just expect it to spit out a vector of split SubSamples. Is there maybe another suggested way to "throw an error" or something like that?

Reply:
Tom, is right ideally we would like to handle errors.

the normal way to throw an error in tooldaq is to return false from from the initialise execute or finalise method that is running, with an associated log statement.

I cant see a situation where you would want to split without the data being sorted, so the sorting proposed by tom is ok.

However another solution would be one could modify your function to have the return type a bool and then the initialise execute or finalise function calling it return false if the function is false. The vector return could then be passed as a referenced argument  variable

Reply:
Assert, whilst ok for purely software situation, when running on the hardware in the DAQ however, we dont want it taking out a whole TPU. A warning and continuing to operate whilst flagging the error is better. Note that in toolDAQ returning a false form initialise execute or finalise, will inform the framework of the failure allowing it to decide on the action to take. It can stop immediately, try to recover, or close the program graciously, running finalise on each tool rather than a hard stop. This lets you disconnect hardware connections etc.

Reply:
I made the `Split` method return an empty vector when things are not sorted. Anyone who calls it can then deal with that however they like. I did not touch `PrepareSubSamples`, because it sorts everything beforehand anyway.